File: aquatrack_app/lib/core/models/intake_log.dart
/// Intake log model for water logging
class IntakeLog {
  final String id;
  final String userId;
  final int volumeMl;
  final String liquidType;
  final double hydrationFactor;
  final int effectiveVolumeMl;
  final int xpEarned;
  final int bonusXp;
  final DateTime loggedAt;
  final DateTime createdAt;
  final String? temperature;
  final String? location;
  final String? moodBefore;
  final String? moodAfter;
  final String source;
  final String? deviceInfo;
  final bool isValidated;
  final double? confidenceScore;

  const IntakeLog({
    required this.id,
    required this.userId,
    required this.volumeMl,
    required this.liquidType,
    required this.hydrationFactor,
    required this.effectiveVolumeMl,
    required this.xpEarned,
    required this.bonusXp,
    required this.loggedAt,
    required this.createdAt,
    this.temperature,
    this.location,
    this.moodBefore,
    this.moodAfter,
    required this.source,
    this.deviceInfo,
    required this.isValidated,
    this.confidenceScore,
  });

  /// Create IntakeLog from JSON
  factory IntakeLog.fromJson(Map<String, dynamic> json) {
    return IntakeLog(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      volumeMl: json['volume_ml'] as int,
      liquidType: json['liquid_type'] as String,
      hydrationFactor: (json['hydration_factor'] as num).toDouble(),
      effectiveVolumeMl: json['effective_volume_ml'] as int,
      xpEarned: json['xp_earned'] as int,
      bonusXp: json['bonus_xp'] as int,
      loggedAt: DateTime.parse(json['logged_at'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
      temperature: json['temperature'] as String?,
      location: json['location'] as String?,
      moodBefore: json['mood_before'] as String?,
      moodAfter: json['mood_after'] as String?,
      source: json['source'] as String,
      deviceInfo: json['device_info'] as String?,
      isValidated: json['is_validated'] as bool,
      confidenceScore: json['confidence_score'] != null
          ? (json['confidence_score'] as num).toDouble()
          : null,
    );
  }

  /// Convert IntakeLog to JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'volume_ml': volumeMl,
      'liquid_type': liquidType,
      'hydration_factor': hydrationFactor,
      'effective_volume_ml': effectiveVolumeMl,
      'xp_earned': xpEarned,
      'bonus_xp': bonusXp,
      'logged_at': loggedAt.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
      'temperature': temperature,
      'location': location,
      'mood_before': moodBefore,
      'mood_after': moodAfter,
      'source': source,
      'device_info': deviceInfo,
      'is_validated': isValidated,
      'confidence_score': confidenceScore,
    };
  }

  /// Get total XP (base + bonus)
  int get totalXp => xpEarned + bonusXp;

  /// Get localized liquid type name
  String get localizedLiquidType {
    switch (liquidType) {
      case 'water':
        return 'Nước lọc';
      case 'tea':
        return 'Trà';
      case 'coffee':
        return 'Cà phê';
      case 'juice':
        return 'Nước trái cây';
      case 'sports_drink':
        return 'Nước thể thao';
      case 'other':
        return 'Khác';
      default:
        return liquidType;
    }
  }

  /// Get localized temperature
  String? get localizedTemperature {
    if (temperature == null) return null;

    switch (temperature!) {
      case 'cold':
        return 'Lạnh';
      case 'room':
        return 'Thường';
      case 'warm':
        return 'Ấm';
      case 'hot':
        return 'Nóng';
      default:
        return temperature;
    }
  }

  /// Create copy with updated fields
  IntakeLog copyWith({
    String? id,
    String? userId,
    int? volumeMl,
    String? liquidType,
    double? hydrationFactor,
    int? effectiveVolumeMl,
    int? xpEarned,
    int? bonusXp,
    DateTime? loggedAt,
    DateTime? createdAt,
    String? temperature,
    String? location,
    String? moodBefore,
    String? moodAfter,
    String? source,
    String? deviceInfo,
    bool? isValidated,
    double? confidenceScore,
  }) {
    return IntakeLog(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      volumeMl: volumeMl ?? this.volumeMl,
      liquidType: liquidType ?? this.liquidType,
      hydrationFactor: hydrationFactor ?? this.hydrationFactor,
      effectiveVolumeMl: effectiveVolumeMl ?? this.effectiveVolumeMl,
      xpEarned: xpEarned ?? this.xpEarned,
      bonusXp: bonusXp ?? this.bonusXp,
      loggedAt: loggedAt ?? this.loggedAt,
      createdAt: createdAt ?? this.createdAt,
      temperature: temperature ?? this.temperature,
      location: location ?? this.location,
      moodBefore: moodBefore ?? this.moodBefore,
      moodAfter: moodAfter ?? this.moodAfter,
      source: source ?? this.source,
      deviceInfo: deviceInfo ?? this.deviceInfo,
      isValidated: isValidated ?? this.isValidated,
      confidenceScore: confidenceScore ?? this.confidenceScore,
    );
  }

  @override
  String toString() =>
      'IntakeLog(id: $id, volume: ${volumeMl}ml, type: $liquidType)';

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is IntakeLog && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;
}

/// Intake log creation request model
class IntakeLogCreateRequest {
  final int volumeMl;
  final String liquidType;
  final String? temperature;
  final String? location;
  final String? moodBefore;
  final String source;

  const IntakeLogCreateRequest({
    required this.volumeMl,
    required this.liquidType,
    this.temperature,
    this.location,
    this.moodBefore,
    required this.source,
  });

  Map<String, dynamic> toJson() {
    return {
      'volume_ml': volumeMl,
      'liquid_type': liquidType,
      if (temperature != null) 'temperature': temperature,
      if (location != null) 'location': location,
      if (moodBefore != null) 'mood_before': moodBefore,
      'source': source,
    };
  }
}

/// Intake log update request model
class IntakeLogUpdateRequest {
  final int? volumeMl;
  final String? liquidType;
  final String? temperature;
  final String? location;
  final String? moodAfter;

  const IntakeLogUpdateRequest({
    this.volumeMl,
    this.liquidType,
    this.temperature,
    this.location,
    this.moodAfter,
  });

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{};

    if (volumeMl != null) json['volume_ml'] = volumeMl;
    if (liquidType != null) json['liquid_type'] = liquidType;
    if (temperature != null) json['temperature'] = temperature;
    if (location != null) json['location'] = location;
    if (moodAfter != null) json['mood_after'] = moodAfter;

    return json;
  }
}

/// Daily summary response model
class DailySummaryResponse {
  final DateTime date;
  final int logCount;
  final int totalVolumeMl;
  final int totalEffectiveMl;
  final int totalXpEarned;

  const DailySummaryResponse({
    required this.date,
    required this.logCount,
    required this.totalVolumeMl,
    required this.totalEffectiveMl,
    required this.totalXpEarned,
  });

  factory DailySummaryResponse.fromJson(Map<String, dynamic> json) {
    return DailySummaryResponse(
      date: DateTime.parse(json['date'] as String),
      logCount: json['log_count'] as int,
      totalVolumeMl: json['total_volume_ml'] as int,
      totalEffectiveMl: json['total_effective_ml'] as int,
      totalXpEarned: json['total_xp_earned'] as int,
    );
  }
}

File: aquatrack_backend/app/crud/intake_log.py
from datetime import date, datetime
from typing import List

from sqlalchemy import and_, func
from sqlalchemy.orm import Session

from app.crud.base import CRUDBase
from app.models.intake_log import IntakeLog
from app.schemas.intake_log import IntakeLogCreate, IntakeLogUpdate


class CRUDIntakeLog(CRUDBase[IntakeLog, IntakeLogCreate, IntakeLogUpdate]):
    """CRUD operations for IntakeLog model"""

    def create(
        self, db: Session, *, obj_in: IntakeLogCreate, user_id: str
    ) -> IntakeLog:
        """Create new intake log with server-side calculations"""
        # Calculate hydration factor based on liquid type
        hydration_factors = {
            "water": 1.0,
            "tea": 0.85,
            "coffee": 0.8,
            "juice": 0.7,
            "sports_drink": 0.9,
            "other": 0.75,
        }

        hydration_factor = hydration_factors.get(obj_in.liquid_type, 0.75)
        effective_volume = int(obj_in.volume_ml * hydration_factor)

        # Flat 20 XP per log, capped at 200 XP per calendar day.
        _XP_PER_LOG = 20
        _DAILY_XP_CAP = 200
        today_xp = (
            db.query(func.coalesce(func.sum(IntakeLog.xp_earned), 0))
            .filter(
                IntakeLog.user_id == user_id,
                func.date(IntakeLog.logged_at) == datetime.now().date(),
            )
            .scalar()
        ) or 0
        base_xp = _XP_PER_LOG if today_xp < _DAILY_XP_CAP else 0

        # Create database object với explicit logged_at
        db_obj = IntakeLog(
            user_id=user_id,
            volume_ml=obj_in.volume_ml,
            liquid_type=obj_in.liquid_type,
            hydration_factor=hydration_factor,
            effective_volume_ml=effective_volume,
            xp_earned=base_xp,
            temperature=obj_in.temperature,
            location=obj_in.location,
            mood_before=obj_in.mood_before,
            source=obj_in.source,
            logged_at=datetime.now(),  # Force current timestamp for consistent date queries
        )

        db.add(db_obj)
        db.commit()
        db.refresh(db_obj)
        return db_obj

    def get_by_user_and_date(
        self,
        db: Session,
        user_id: str,
        target_date: date,
    ) -> List[IntakeLog]:
        """Get all intake logs for a user on a specific date"""
        return (
            db.query(self.model)
            .filter(
                and_(
                    self.model.user_id == user_id,
                    func.date(self.model.logged_at) == target_date,
                )
            )
            .order_by(self.model.logged_at.desc())
            .all()
        )

    def get_by_user_date_range(
        self,
        db: Session,
        user_id: str,
        date_from: date,
        date_to: date,
        skip: int = 0,
        limit: int = 100,
    ) -> List[IntakeLog]:
        """Get intake logs for a user within a date range"""
        return (
            db.query(self.model)
            .filter(
                and_(
                    self.model.user_id == user_id,
                    func.date(self.model.logged_at) >= date_from,
                    func.date(self.model.logged_at) <= date_to,
                )
            )
            .order_by(self.model.logged_at.desc())
            .offset(skip)
            .limit(limit)
            .all()
        )

    def get_recent_by_user(
        self,
        db: Session,
        user_id: str,
        limit: int = 10,
    ) -> List[IntakeLog]:
        """Get recent intake logs for a user"""
        return (
            db.query(self.model)
            .filter(self.model.user_id == user_id)
            .order_by(self.model.logged_at.desc())
            .limit(limit)
            .all()
        )

    def get_daily_stats(
        self,
        db: Session,
        user_id: str,
        target_date: date,
    ) -> dict:
        """Get daily intake statistics for a user"""
        result = (
            db.query(
                func.count(self.model.id).label("log_count"),
                func.coalesce(func.sum(self.model.volume_ml), 0).label("total_volume"),
                func.coalesce(func.sum(self.model.effective_volume_ml), 0).label(
                    "total_effective"
                ),
                func.coalesce(
                    func.sum(self.model.xp_earned + self.model.bonus_xp), 0
                ).label("total_xp"),
                func.avg(self.model.volume_ml).label("avg_volume"),
            )
            .filter(
                and_(
                    self.model.user_id == user_id,
                    func.date(self.model.logged_at) == target_date,
                )
            )
            .first()
        )

        return {
            "date": target_date,
            "log_count": result.log_count or 0,
            "total_volume_ml": result.total_volume or 0,
            "total_effective_ml": result.total_effective or 0,
            "total_xp_earned": result.total_xp or 0,
            "average_volume_ml": float(result.avg_volume or 0),
        }

    def get_liquid_type_stats(
        self,
        db: Session,
        user_id: str,
        date_from: date,
        date_to: date,
    ) -> List[dict]:
        """Get liquid type distribution statistics"""
        result = (
            db.query(
                self.model.liquid_type,
                func.count(self.model.id).label("count"),
                func.sum(self.model.volume_ml).label("total_volume"),
                func.sum(self.model.effective_volume_ml).label("total_effective"),
                func.avg(self.model.volume_ml).label("avg_volume"),
            )
            .filter(
                and_(
                    self.model.user_id == user_id,
                    func.date(self.model.logged_at) >= date_from,
                    func.date(self.model.logged_at) <= date_to,
                )
            )
            .group_by(self.model.liquid_type)
            .all()
        )

        return [
            {
                "liquid_type": r.liquid_type,
                "log_count": r.count,
                "total_volume_ml": r.total_volume or 0,
                "total_effective_ml": r.total_effective or 0,
                "average_volume_ml": float(r.avg_volume or 0),
            }
            for r in result
        ]

    def count_by_user_and_date(
        self,
        db: Session,
        user_id: str,
        target_date: date,
    ) -> int:
        """Count intake logs for a user on a specific date"""
        return (
            db.query(func.count(self.model.id))
            .filter(
                and_(
                    self.model.user_id == user_id,
                    func.date(self.model.logged_at) == target_date,
                )
            )
            .scalar()
        )


def authoritative_total_xp(db: Session, user) -> int:
    """The user's true Total XP — the single value the Level is derived from
    everywhere (matches /levels/current): repeatable intake XP
    (xp_earned + bonus_xp) plus the quest/milestone XP stored on user.total_xp.
    Use this (never user.total_xp alone) when reconciling Level-Up coins so the
    high-water mark tracks the real level."""
    intake_xp = (
        db.query(func.sum(IntakeLog.xp_earned + IntakeLog.bonus_xp))
        .filter(IntakeLog.user_id == user.id)
        .scalar()
        or 0
    )
    return int(intake_xp) + (user.total_xp or 0)


# Global instance
intake_log_crud = CRUDIntakeLog(IntakeLog)
